### ISSUE TITLE
2.2 Return RedirectToPage from Post Handler issues

### ISSUE THREAD

**Issue by robertburgh** [aspnet/AspNetCore.Docs#13542]
Using the same syntax of return RedirectToPage("/Index"); inside of a post handler doesn't appear to cause the browser to accept the new url.
The output window in VS shows the complete and correct full url (that can be copied and pasted directly into the browser with expected results).
I tried it with and without [IgnoreAntiforgeryToken] on the target page.
Thanks

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 164d7efe-01bf-0b26-7a23-f10704f338f5
* Version Independent ID: b3e2bc87-1582-2d34-c7c2-7c8e901ac771
* Content: [Introduction to Razor Pages in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/razor-pages/?view=aspnetcore-2.2&tabs=visual-studio#feedback)
* Content Source: [aspnetcore/razor-pages/index.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/razor-pages/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-razorpages**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
I'm not sure what's going on but `RedirectToPage` does work.

**Comment by Rick-Anderson** [comment]
`return RedirectToPage("./Index");` is used in the intro to RP and the EF/RP series.

**Comment by robertburgh** [comment]
1.) I have tried with "./" and "/" multiple times.
2.) Breaking on the OnPost shows that is it firing as expected.
3.) The Visual Studio (2017) Output Window shows the expected url being created.

I created new projects to test this issue and am getting the same unexpected results.

There are no errors in the Browser or VS.

In the .cs file:
        public void OnGet()
        {
        }
        public IActionResult OnPostTest()
        {
            return RedirectToPage("./Privacy");
        }
In the cshtml, I am using ajax for the post. This mimics the basic need of the real project that is beyond the scope of this issue.
For this test, I am using a button and onclick
    function clickThisButton()
    {
        var url = '/Index?handler=Test'
        console.log(url);
        console.log($('input[name=__RequestVerificationToken]').val());
        $.ajax({
            url: url,
            headers: { "RequestVerificationToken": $('input[name="__RequestVerificationToken"]').val() },
            type: 'POST'
        });
    }
Is there anything that recent updates may have affected the behavior of RedirectToPage?
Is there anything that might impede the function of RedirectToPage? Since the VS output shows the expected url, the brower's not updating to the update url is curious.

**Comment by robertburgh** [comment]
This is from the Output of Visual Studio when the Button is clicked that fires the ajax Post:

This shows that "Executed endpoint 'Page: /Test'" but the browser never leaves /Index.




Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 POST https://localhost:44317/Index?handler=Test  0
Microsoft.AspNetCore.Routing.EndpointMiddleware:Information: Executing endpoint 'Page: /Index'
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Route matched with {page = "/Index"}. Executing page /Index
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executing handler method zRazorPageRedirect.Pages.IndexModel.OnPostTest with arguments ((null)) - ModelState is Valid
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executed handler method OnPostTest, returned result Microsoft.AspNetCore.Mvc.RedirectToPageResult.
Microsoft.AspNetCore.Mvc.RedirectToRouteResult:Information: Executing RedirectToPageResult, redirecting to ./Test.
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executed page /Index in 4.5391ms
Microsoft.AspNetCore.Routing.EndpointMiddleware:Information: Executed endpoint 'Page: /Index'
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 19.0664ms 302 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET https://localhost:44317/Test  
Microsoft.AspNetCore.Routing.EndpointMiddleware:Information: Executing endpoint 'Page: /Test'
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Route matched with {page = "/Test"}. Executing page /Test
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executing handler method zRazorPageRedirect.Pages.TestModel.OnGet with arguments ((null)) - ModelState is Valid
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executed handler method OnGet, returned result .
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executing an implicit handler method - ModelState is Valid
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executed an implicit handler method, returned result Microsoft.AspNetCore.Mvc.RazorPages.PageResult.
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker:Information: Executed page /Test in 7.7897ms
Microsoft.AspNetCore.Routing.EndpointMiddleware:Information: Executed endpoint 'Page: /Test'
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 29.8774ms 200 text/html; charset=utf-8

**Comment by serpent5** [comment]
The difference between the docs examples, etc, is that they're not using AJAX. Redirects don't work the same way for AJAX as they do for regular browser requests.

**Comment by robertburgh** [comment]
> The difference between your example and the docs examples, etc, is that they're not using AJAX. Redirects don't work the same way for AJAX as they do for regular browser requests.

@serpent5 I thank you for the response. I can't find any documentation citing that. Can you point me in the right direction.

If I understand what you are saying, it is because the ajax POST fires the handler that contains the RedirectToPage, then it middleware doesn't get the response back to the browser?

**Comment by serpent5** [comment]
I don't know if there's any real documentation for that: it's somewhat fundamental to how HTTP and browsers work, so it's certainly not specific to ASP.NET Core. I suggest you add a `success` handler to your jQuery call and inspect the response - that might help you learn about what's really going on.

**Comment by brcaswell** [comment]
It's not really accurate to say this is a matter of HTTP protocol or browser behavior.  You can do a 302 response with AJAX and browsers support the redirect to location header (given domain and CORS).  there are in fact framework policies that distinguish and handle response statuscode conditionally on whether the request is an AJAX request (in OWIN Katana specifically) based on intent.  In the matter of unauthorized responses with middleware, ApiController (WebAPI) handles AJAX requests with a 401 and Controller (MVC) handles it with a 302 redirect; So this matter may be something to consider and covered by this introduced MVC pattern that doesn't use MVC Controller.

**Comment by SoftCircuits** [comment]
@robertburgh Did you ever learn anything about this? I see the issue was closed but I don't see any resolution. I'm looking into redirecting from the AJAX handler too and it won't work for me under .NET Core 3.1 Razor Pages. But when I search GitHub repositories, it seems there's a lot of code doing exactly this.

**Comment by robertburgh** [comment]
> @robertburgh Did you ever learn anything about this? I see the issue was closed but I don't see any resolution. I'm looking into redirecting from the AJAX handler too and it won't work for me under .NET Core 3.1 Razor Pages. But when I search GitHub repositories, it seems there's a lot of code doing exactly this.

@SoftCircuits  What @Rick-Anderson says about the redirects working is accurate. For my issues, it wasn't about ajax, the POST verb, etc. but troubleshooting several - in hindsight - minor issues. None of which had anything to do directly with redirect. Example: Antiforgery token, ./ vs /  and other small issues. I disabled everything I could, fired up postman, and most important remembered that what I knew was still valid even if I wasn't yet comfortable with razor pages.

Bottom line. Troubleshoot using postman or the tool of your choice. It is going to be something or things that are simple. If the post fires, it will redirect. It doesn't matter what causes the post to fire (helper tags, ajax, etc.)

I fully get that the folks building and maintaining .net core aren't in the business in reminding us about using basic troubleshooting. Perhaps they forgot what it is like to not be sure how to troubleshoot .net core? That aside, I do appreciate the work they are doing and I think razor pages are great.
